clash-ghc/src-ghc/Clash/GHC/LoadInterfaceFiles.hs

```diff
@@ -168,145 +256,132 @@ loadIface foundMod = do
                                            ]
                          in traceIf True msg' (return Nothing)
 
-loadExternalBinders
-  :: GHC.GhcMonad m
-  => HDL
-  -> [CoreSyn.CoreBndr]
-  -> m LoadedBinders
+-- | Given a list of top-level binders, recursively load all the binders,
+-- primitives, and type classes it is using. (Exported function.)
+loadExternalBinders :: GHC.GhcMonad m => HDL -> [CoreSyn.CoreBndr] -> m LoadedBinders
 loadExternalBinders hdl bndrs = do
-  loaded <- mergeLoadedBinders <$> mapM (loadExprFromIface hdl) bndrs
-  fst <$>
-    loadExternalExprs'
-      hdl
-      loaded
-      (UniqSet.mkUniqSet (collectLbBinders loaded))
-      (map snd (lbBinders loaded))
-
-loadExternalExprs
-  :: GHC.GhcMonad m
-  => HDL
-  -> UniqSet.UniqSet CoreSyn.CoreBndr
-  -> [CoreSyn.CoreBind]
-  -> m LoadedBinders
-loadExternalExprs hdl = go emptyLb
-  where
-    go loaded _ [] =
-      return loaded
-    go loaded0 visited0 (CoreSyn.NonRec _ e:bs) = do
-      (loaded1, visited1) <- loadExternalExprs' hdl loaded0 visited0 [e]
-      go loaded1 visited1 bs
-    go loaded0 visited0 (CoreSyn.Rec bs:bs') = do
-      (loaded1, visited1) <- loadExternalExprs' hdl loaded0 visited0 (map snd bs)
-      go loaded1 visited1 bs'
-
--- | Used by entry points: 'loadExternalExprs', 'loadExternalBinders'
-loadExternalExprs'
-  :: GHC.GhcMonad m
-  => HDL
-  -> LoadedBinders
-  -> UniqSet.UniqSet CoreSyn.CoreBndr
-  -> [CoreSyn.CoreExpr]
-  -> m ( LoadedBinders, UniqSet.UniqSet CoreSyn.CoreBndr)
-loadExternalExprs' _hdl loaded visited [] =
-  return (loaded, visited)
-loadExternalExprs' hdl loaded0 visited0 (e:es) = do
-  let
-    isInteresting v =
-         Var.isId v
-      && not (v `UniqSet.elementOfUniqSet` visited0)
-      && isNothing (Id.isDataConId_maybe v)
-
-    fvs0 = CoreFVs.exprSomeFreeVarsList isInteresting e
-    fvs1 = map (\v -> maybe (Left v) (Right . (v,)) (Id.isClassOpId_maybe v)) fvs0
-    (fvs2, clsOps0) = partitionEithers fvs1
-    clsOps1 = map goClsOp clsOps0
-
-  loaded1 <- mergeLoadedBinders <$> mapM (loadExprFromIface hdl) fvs2
-
-  loadExternalExprs'
-    hdl
-    (mergeLoadedBinders [loaded0, loaded1, emptyLb{lbClassOps=clsOps1}])
-    (foldl' UniqSet.addListToUniqSet visited0 [collectLbBinders loaded1, map fst clsOps0])
-    (es ++ map snd (lbBinders loaded1))
+  flip execStateT emptyLb $ do
+    mapM_ (loadExprFromIface hdl) bndrs
+
+-- Given a list of binds, recursively load all its binders, primitives, and
+-- type classes it is using. (Exported function.)
+loadExternalExprs :: GHC.GhcMonad m => HDL -> [CoreSyn.CoreBind] -> m LoadedBinders
```

Comment:
Given the [recent mention of loading primitives later](https://github.com/clash-lang/clash-compiler/issues/2020#issuecomment-989669791) and the fact this part of the code base is about loading GHC things, it feels strange to me on re-read that `HDL` is given as an argument to these things. I think we should open an issue around

  1. loading primitives later in the compilation process (ideally in `clash-lib` not `clash-ghc`, since there is nothing inherently GHC-specific about loading Clash primitives)
  2. Removing the `HDL` parameter from functions that load haskell modules / interfaces etc.

Reply:
Agreed. I'll open an issue, but it's not something I want addressed in this PR (as you also imply, I think).

Reply:
The argument is provided because we load the primitives from `ANN` pragma's and it could be the case the primitive is only defined for a certain language.

There are reasons that primitives could only specified for certain backends
1. As a work-around for a certain backend
2. You have some custom logic and you (a non-clash-core-developer but just an advanced user) only care about one language

Reply:
Filtering could still happen after binder loading though. That way you could potentially make this backend agnostic and cache the process (however useless that may be). In any case I don't thing it's too important.

Reply:
I needs to happen in `clash-ghc` because that's where `GHC2Core` lives, which uses the information to determine whether to turn a GHC Core `Var` into a Clash Core `Var` or `Prim`.